### PR TITLE
feat: add spot termination handler

### PR DESCRIPTION
**Pull request by npalm**
## Description

This PR adds a lambda to to log and metric spot termination based on the cloudtrail event `BidEvictedEvent`. The feature is experimental and disabled by default.

## Future directions

The current implemenation only helps to make spot termination visible to an admin team. For the future we want to enrich a runner with information via tagging what job is active, This allows to let the termination handler also inform the user by adding a job annotation once a spot termination occurs.

## Migration

No migration is required. By default the watcher is disabled.

- logging for the watcher is changed
- resources will be recreated for notification warning watcher

### CODE REVIEW

**Review by mpas — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -165,7 +165,7 @@ Talk to the forestkeepers in the `runners-channel` on Slack.
...
...re is Beta, changes will not trigger a major release as long in beta.<br/><br/>`enable`: Enable or disable the spot termination watcher.<br/>'features': Enable or disable features of the termination watcher.<br/>`memory_size`: Memory size linit in MB of the lambda.<br/>`s3_key`: S3 key for syncer lambda function. Required if using S3 bucket to specify lambdas.<br/>`s3_object_version`: S3 object version for syncer lambda function. Useful if S3 versioning is enabled on source bucket.<br/>`timeout`: Time out of the lambda in seconds.<br/>`zip`: File location of the lambda zip file. | <pre>object({<br/>    enable        = optional(bool, false)<br/>    enable_metric = optional(string, null) # deprectaed<br/>    features = optional(object({<br/>      enable_spot_termination_handler              = optional(bool, true)<br/>      enable_spot_termination_notification_watcher = optional(bool, true)<br/>    }), {})<br/>    memory_size       = optional(number, null)<br/>    s3_key            = optional(string, null)<br/>    s3_object_version = optional(string, null)<br/>    timeout           = optional(number, null)<br/>    zip               = optional(string, null)<br/>  })</pre> | `{}` | no |
```
Comment (stuartp44) [inline README.md:168]:
```suggestion
| <a name="input_instance_termination_watcher"></a> [instance\_termination\_watcher](#input\_instance\_termination\_watcher) | Configuration for the instance termination watcher. This feature is Beta, changes will not trigger a major release as long in beta.<br/><br/>`enable`: Enable or disable the spot termination watcher.<br/>'features': Enable or disable features of the termination watcher.<br/>`memory_size`: Memory size linit in MB of the lambda.<br/>`s3_key`: S3 key for syncer lambda function. Required if using S3 bucket to specify lambdas.<br/>`s3_object_version`: S3 object version for syncer lambda function. Useful if S3 versioning is enabled on source bucket.<br/>`timeout`: Timeout of the lambda in seconds.<br/>`zip`: File location of the lambda zip file. | <pre>object({<br/>    enable        = optional(bool, false)<br/>    enable_metric = optional(string, null) # deprectaed<br/>    features = optional(object({<br/>      enable_spot_termination_handler              = optional(bool, true)<br/>      enable_spot_termination_notification_watcher = optional(bool, true)<br/>    }), {})<br/>    memory_size       = optional(number, null)<br/>    s3_key            = optional(string, null)<br/>    s3_object_version = optional(string, null)<br/>    timeout           = optional(number, null)<br/>    zip               = optional(string, null)<br/>  })</pre> | `{}` | no |
```

File: docs/configuration.md [hunk 2]
```diff
@@ -215,21 +215,35 @@ In case the setup does not work as intended, trace the events through this seque
 
 ### Termination watcher
 
-This feature is in early stage and therefore disabled by default.
+This feature is in early stage and therefore disabled by default. To enable the watcher set `instance_termination_watcher.enable = true`.
```
Comment (stuartp44) [inline docs/configuration.md:218]:
```suggestion
This feature is in early stage and therefore disabled by default. To enable the watcher, set `instance_termination_watcher.enable = true`.
```

File: docs/configuration.md [hunk 3]
```diff
@@ -215,21 +215,35 @@ In case the setup does not work as intended, trace the events through this seque
...
 
-The termination watcher is currently watching for spot termination notifications. The module is only taken events into account for instances tagged with `ghr:environment` by default when deployment the module as part of one of the main modules (root or multi-runner). The module can also be deployed stand-alone, in that case the tag filter needs to be tunned.
+The termination watcher is currently watching for spot terminations. The module is only taken events into account for instances tagged with `ghr:environment` by default when deployment the module as part of one of the main modules (root or multi-runner). The module can also be deployed stand-alone, in that case the tag filter needs to be tunned.
```
Comment (stuartp44) [inline docs/configuration.md:220]:
```suggestion
The termination watcher is currently watching for spot terminations. The module is only taken events into account for instances tagged with `ghr:environment` by default when deployment the module as part of one of the main modules (root or multi-runner). The module can also be deployed stand-alone, in this case, the tag filter needs to be tunned.
```

File: examples/default/main.tf [hunk 4]
```diff
@@ -78,7 +78,7 @@ module "runners" {
   # Let the module manage the service linked role
   # create_service_linked_role_spot = true
 
-  instance_types = ["m5.large", "c5.large"]
+  instance_types = ["m7a.large", "m5.large"]
```
Comment (stuartp44) [inline examples/default/main.tf:81]: I dont think this is a intended change unless its a piggyback

File: modules/multi-runner/README.md [hunk 5]
```diff
@@ -136,7 +136,7 @@ module "multi-runner" {
...
...guration for the spot termination watcher lambda function. This feature is Beta, changes will not trigger a major release as long in beta.<br/><br/>`enable`: Enable or disable the spot termination watcher.<br/>`memory_size`: Memory size linit in MB of the lambda.<br/>`s3_key`: S3 key for syncer lambda function. Required if using S3 bucket to specify lambdas.<br/>`s3_object_version`: S3 object version for syncer lambda function. Useful if S3 versioning is enabled on source bucket.<br/>`timeout`: Time out of the lambda in seconds.<br/>`zip`: File location of the lambda zip file. | <pre>object({<br/>    enable         = optional(bool, false)<br/>    enable_metrics = optional(string, null) # deprecated<br/>    features = optional(object({<br/>      enable_spot_termination_handler              = optional(bool, true)<br/>      enable_spot_termination_notification_watcher = optional(bool, true)<br/>    }), {})<br/>    memory_size       = optional(number, null)<br/>    s3_key            = optional(string, null)<br/>    s3_object_version = optional(string, null)<br/>    timeout           = optional(number, null)<br/>    zip               = optional(string, null)<br/>  })</pre> | `{}` | no |
```
Comment (stuartp44) [inline modules/multi-runner/README.md:139]:
```suggestion
| <a name="input_instance_termination_watcher"></a> [instance\_termination\_watcher](#input\_instance\_termination\_watcher) | Configuration for the spot termination watcher lambda function. This feature is Beta, changes will not trigger a major release as long as its in beta.<br/><br/>`enable`: Enable or disable the spot termination watcher.<br/>`memory_size`: Memory size linit in MB of the lambda.<br/>`s3_key`: S3 key for syncer lambda function. Required if using S3 bucket to specify lambdas.<br/>`s3_object_version`: S3 object version for syncer lambda function. Useful if S3 versioning is enabled on source bucket.<br/>`timeout`: Time out of the lambda in seconds.<br/>`zip`: File location of the lambda zip file. | <pre>object({<br/>    enable         = optional(bool, false)<br/>    enable_metrics = optional(string, null) # deprecated<br/>    features = optional(object({<br/>      enable_spot_termination_handler              = optional(bool, true)<br/>      enable_spot_termination_notification_watcher = optional(bool, true)<br/>    }), {})<br/>    memory_size       = optional(number, null)<br/>    s3_key            = optional(string, null)<br/>    s3_object_version = optional(string, null)<br/>    timeout           = optional(number, null)<br/>    zip               = optional(string, null)<br/>  })</pre> | `{}` | no |
```

File: examples/default/main.tf (hunk 4 above)
Comment (npalm) [reply examples/default/main.tf:81]: It is intended, and it is just an example. Got too many issues with c5

**Review by stuartp44 — APPROVED** [review]

**Pull request merged**